### ISSUE TITLE
build fails on Windows

### ISSUE THREAD

**Issue by Vandivier** [backspaces/asx#13]
`npm run build` invokes the clean task which invokes command `rm`. This is not a valid Windows CLI command. The equivalent would be `del`

Console:
'rm' is not recognized as an internal or external command...

ref: https://stackoverflow.com/questions/33188752/rm-is-not-recognized-as-an-internal-or-external-command-while-using-meteor
ref: https://github.com/meanjs/generator-meanjs/issues/121

**Comment by Vandivier** [comment]
also the back ticks, designed to evaluate an expression in-place, also fail on windows.

I tried Git Bash thinking bash might do the trick but no such luck

I'm rewriting the build steps as actual Node and I'll submit a PR

**Comment by backspaces** [comment]
I don't have access to a windows machine, and my team either use Mac/Linux or Cygwin.

I did have it using shelljs/shx: https://github.com/shelljs/shx for a while but then decided it would be unnecessary.

Could you give a look at shx? I'd prefer something like that rather than using raw node for shell level commands and it would be familiar to mac/linux/cygwin users.

Thanks for looking into this.

**Comment by backspaces** [comment]
BTW: these may help understand the avoidance of the typical workflow tools and the dual build approach:

> https://medium.com/@backspaces/es6-modules-part-1-migration-strategy-a48de0b7f112
https://medium.com/@backspaces/es6-modules-part-2-libs-wrap-em-up-8715e116d690

**Comment by Vandivier** [comment]
shx looks great. I'll give it a go this evening or next.

**Comment by backspaces** [comment]
Yay! It'd be great to be windows-friendly.

**Comment by Vandivier** [comment]
I'm running into the following issue: `bin/pkgkey.js foo` cannot be evaluated in-place.

This seems to work in Git Bash: echo `bin/pkgkey.js mkdirs`

But, given a script entry "ech": "echo `bin/pkgkey.js mkdirs`", this fails:
`npm run ech`

Instead, it prints to cli `bin/pkgkey.js mkdirs`

shx does not improve this. "ech": "shx echo `bin/pkgkey.js mkdirs`" falls into the same issue.

This is an issue in many of the scripts to be converted. Eg "mkdirs": "mkdir -p `bin/pkgkey.js mkdirs`"

I am able to workaround by separately calling mkdir -p directly on each entry into pkgkey.js without using pkgkey.js

So, we can workaround with shx but we lose maintainability. How should I proceed?

I can keep a separate build-windows so there is no conflict. But, I could do that with the all-node approach as well.

**Comment by Vandivier** [comment]
Piping also seems to fail: `bin/pkgkey.js mkdirs | mkdir -p`

**Comment by backspaces** [comment]
Thanks for hanging in there.

So basically backticks: 
> mkdir -p \`bin/pkgkey.js mkdirs`

and piping:
> bin/pkgkey.js | mkdir -p

fail, right? Nice idea bout piping, by the way!

So the issue it how to make data in the package.json available to run scripts, and our current stunt with pkgkey.js is not cross platform.

I did check pkgkey.js and it does use standard out so piping should work for commands accepting stdin. But the places we use pkgkey:
```
rm -rf `bin/pkgkey.js mkdirs`
mkdir -p `bin/pkgkey.js mkdirs`
cp `bin/pkgkey.js libs` libs
cp -Rp `bin/pkgkey.js docsfiles` docs
```
.. don't accept stdin. Sigh

Lets think about this for a bit, I'm off to a meeting. At worst, we can use node scripts for these four. Or maybe make a few general scripts for rm, mkdir, cp that know to use the pkg key for their first argument: cpkey -rf 'mkdirs' and so on.

**Comment by Vandivier** [comment]
Yes, that is the issue. Happy to take our time and think about it.

For now I'm pressing on with scripts but happy to revisit later. I'm trying to use unix-like functions where possible including shellJs. Take a look:
https://github.com/Vandivier/charm-abm/blob/master/js/build.js

**Comment by backspaces** [comment]
Here's a Stack Overflow question on this:
https://stackoverflow.com/questions/46208615/how-use-backticks-in-npm-scripts-on-windows

**Comment by Vandivier** [comment]
PR #14 submitted

**Comment by backspaces** [comment]
See PR .. I have submitted an npm bug: https://github.com/npm/npm/issues/18580

**Comment by backspaces** [comment]
I've created a separate issue for the wonky rollup.config: #15

**Comment by backspaces** [comment]
Here's more on npm's shell. 
> https://stackoverflow.com/questions/23243353/how-to-set-shell-for-npm-run-scripts-in-windows
https://github.com/npm/npm/issues/9420

If we can specify the shell used by npm, then we should be able to solve the cross-platform problem by using the same shell.

**Comment by backspaces** [comment]
Check #15 for interesting idea for eliminating external/global modules from bundle.
Also, just checked in removing bin/wraplibplus.js which simplifies workflow a bit.

**Comment by backspaces** [comment]
Updating rollup.config should solve problem mentioned above. Removing wraplibplus should reduce windows issues. See recent commits.

**Comment by backspaces** [comment]
I'm going to close this for now. We have a new architect on board and I believe we'll go to a build system like webpack that may be cross platform.

**Comment by backspaces** [comment]
@Vandivier I'm reopening this. Why? Our team is converting to CircleCI's Continuous Integration system. It includes using docker which will build in a standard environment, a ubuntu unix environment.

I know about zero about this, but I think this would allow a cross platform build system. The main issue is that it is designed for micro commits to github, with another server running tests etc. I don't know how awkward this would be.

Bottom line: a potential way to have windows builds that is super standard and part of team workflows just might also solve this issue.

**Comment by Vandivier** [comment]
you just said so many things i like

**Comment by backspaces** [comment]
@Vandivier Could you look at https://github.com/backspaces/asx/pull/32?

**Comment by Vandivier** [comment]
Will do this PM

**Comment by Vandivier** [comment]
environment:
1. Win10x64, git bash
2. node -v = 9.3.0
3. npm -v = 4.0.5

steps taken:
1. Checked out the latest ASX as of 7 AM today, 2/1
2. Manually added package.json changes from `gucong3000:x-shell`, `77b80fd`
3. `npm install`
4. `npm build`
5. `npm run build`
6. modify to prepend `bin`

findings:
1. node-sass@3.13.1 postinstall fails during step 3. I seem to recall this isn't critical.
2. `npm build` returns instantaneously without doing anything.
3.  `npm run build` throws no entity error (ENOENT): "/c/Program: Files\Git\usr\bin\bash: No such file or directory". This appears to be due to unqualified `bin\pkgkey.js` and similar.
4. Prepending `bin` with any of the following doesn't resolve path issue:
  1. `/`
  2. `./`
  3. `${BASH_SOURCE[0]}/`

image below of steps 4 and 5:

![image](https://user-images.githubusercontent.com/5559355/35678886-120c4358-0723-11e8-90be-d2e22b887db0.png)

**Comment by backspaces** [comment]
It turns out that I'm working on asx such that the windows build may become easier.
- I'm upgrading to all-ESM (es6 modules) for the three dependencies: Three, dat.gui, Stats.js 
  - I'm pulling the last two out of asx, they were conveniences building the Three view but poor modularity.
  - So only three.js is left and I'm going to use the ESM version
  - This removes a lot of workflow, no "wrapping" of iife's to modules.
- I can remove the backticked pkgkey.js utility.
  - I can replace it with env vars. There is a windows `cross-env` solution for that.
  - Or just not bother
- I will probably use shelljs/shx for existing bash scripts. It is already used in my new test/ world.

I'm not sure how deep a dive this is, but when I get to a place where it needs testing, I'll post here and on the PR by @gucong3000. I do *love* the idea of an absolutely cross platform bash tho.

**Comment by Vandivier** [comment]
Great! FWIW I've used `cross-env` to success a number of times.

**Comment by backspaces** [comment]
BTW: the issue I submitted to npm: https://github.com/npm/npm/issues/18580 could be solved with  @gucong3000's git-bash-shell. Possibly we could talk the npm team into making it a core bash for windows?

**Comment by backspaces** [comment]
I've made my first commit that helps clean up some of the problems. I used env vars (those within package.json itself) in two places. So I'll convert those to use `cross-env`. 

I also used shx in one npm script (rm -rf .. rimraf equivalent) and have started using it in the couple of node scripts I have.

An as @gucong3000 suggested, I converted to npm-run-all .. a lot cleaner.

**Comment by backspaces** [comment]
I've made a second pass, removing all bash scripts from bin/ .. the build utilities. There are now just 3 node utilities.

Question: when I call my node utilities, I just do this:
>    "build-libs": "bin/wraplibs.js",

.. because the node file has a "shebang": 
>    #!/usr/bin/env node

Will I have to convert to calling the utilities like this?
>    "build-libs": "node bin/wraplibs.js",

One issue: I'm afraid that the tests may not be made cross-platform currently. I use ava to launch Puppeteer (chrome's headless browser interface) to run all the demo models for 4 seconds, making screen grab images to see if they work and if they are the same as earlier.

But this really isn't a huge deal because the as-core repo is the one whose tests are most important and I believe I can make them cross-platform.

**Comment by backspaces** [comment]
OK, I've checked in the latest attempt to be cross-platform.

Could someone try a new clone, yarn, yarn run build?

It is just possible the tests can work, they're written in AvA but use Puppeteer which should install the correct chromium but I can't try it. Just `yarn test`

**Comment by backspaces** [comment]
@Vandivier could you try clone/install/build of asx? I think it's pretty close to windows usable.

**Comment by backspaces** [comment]
Hi @Vandivier .. thanks for the help!

I saw this: https://github.com/backspaces/asx/pull/32#issuecomment-364831275 which lead me to believe you tried the clone/yarn/yarn build on the PR's repo, right? (I'm not trying his portable bash approach, much too heavy and system dependent. Altho npm itself should consider it.)

Could you try this instead?
```
git clone https://github.com/backspaces/asx
cd asx
yarn
yarn build
```

I'm hoping this will yield a cross-platform build. Or at least be closer! There are *NO* bash scripts.

One warning: I'm now starting testing, and am using Puppeteer. It's install is sometimes flaky. Just delete it from the devDependencies if it is a problem. Only used in test. 

But if the build *does* work, you might `yarn test`. It probably is not cross-platform but if it is, you'll like what happens.

**Comment by backspaces** [comment]
@Vandivier No need to test. Looks like we get the same error on a windows system. Three has been a nightmare! Thanks for the help.

I think I'm going to have to abandon the "wrapper" trick and use three.module.js with examples/*js files run through a script to make it work.

![img_20180212_154247_1024](https://user-images.githubusercontent.com/124366/36124164-4533eb7c-100c-11e8-8073-61164676a59f.jpg)

**Comment by backspaces** [comment]
We're getting close!

We solved the problem above which was a shebang issue: I changed this:
> exec(`bin/wraplib.js ${fromPath} ${name} > ${toPath}`)

to this:
> exec(`node bin/wraplib.js ${fromPath} ${name} > ${toPath}`)

This solved the problem of wrapper files being empty!

But now we still fail in rollup with the same error. And we "fix" it by changing the source path to use "\" rather than "/". I believe this is a rollup bug. From my team mate:

>I get a successful build if I change the slashes in `ThreeMeshes.js` and `ThreeView.js` for the dist imports from `/` to `\`

Do any of us have a similar experience? Possibly a fix?

**Comment by backspaces** [comment]
OK, I've submitted an Issue on this to Rollup:
https://github.com/rollup/rollup/issues/1967

Note this problem even if bash were available on both systems.
